src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/TextDecorationCollectionConverter.cs

```diff
@@ -2,285 +2,159 @@
-        /// <param name="destinationType"> The destination type of the conversion </param>
-        /// <returns> null will always be returned because TextDecorations cannot be converted to any other type. </returns>        
-        public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
-        {
-            if (destinationType == typeof(InstanceDescriptor) && value is IEnumerable<TextDecoration>)
+            // Go through each item in the input and match accordingly
+            foreach (Range segment in decorationsSpan.Split(','))
             {
-                ConstructorInfo ci = typeof(TextDecorationCollection).GetConstructor(
-                    new Type[]{typeof(IEnumerable<TextDecoration>)}
-                    );
-                    
-                return new InstanceDescriptor(ci, new object[]{value});                
-            }
-
-            // Pass unhandled cases to base class (which will throw exceptions for null value or destinationType.)
-            return base.ConvertTo(context, culture, value, destinationType);              
-        }
-
-        //---------------------------------
-        // Private methods
-        //---------------------------------
-        /// <summary>
-        /// Match the input against a predefined pattern from certain index onwards
-        /// </summary>
-        private static bool Match(string pattern, string input, int index)
-        {
-            int i = 0;
-            for (;
-                    i < pattern.Length
-                 && index + i < input.Length
-                 && pattern[i] == Char.ToUpperInvariant(input[index + i]);
-                 i++) ;
+                ReadOnlySpan<char> decoration = decorationsSpan[segment].Trim();
 
-            return (i == pattern.Length);
-        }
-
-        /// <summary>
-        /// Advance to the start of next name
-        /// </summary>
-        private static int AdvanceToNextNameStart(string input, int index)
-        {
-            // Two names must be seperated by a comma and optionally spaces
-            int separator = AdvanceToNextNonWhiteSpace(input, index);
-
-            int nextNameStart;
-            if (separator >= input.Length)
-            {
-                // reach the end
-                nextNameStart = input.Length;
-            }
-            else
-            {
-                if (input[separator] == Separator)
-                {                    
-                    nextNameStart = AdvanceToNextNonWhiteSpace(input, separator + 1);
-                    if (nextNameStart >= input.Length)
-                    {
-                        // Error: Separator is at the end of the input
-                        nextNameStart = -1;
-                    }
+                if (decoration.Equals("Overline", StringComparison.OrdinalIgnoreCase) && (matchedDecorations & OverlineMatch) == 0)
```

Comment:
Since this would be a case when we're already going to throw if it doesn't work out, the other day @miloush told me I don't need to optimize anymore; comparing the text first seemed more readable as you get it aligned under eachother.

(miloush, what is that thumbs up doing there?!)

Reply:
Thumbs up was for doing cheaper conditions first in general. Doesn't this throw only when all the conditions fail?

Reply:
I can agree on that too 😀 So let's take a look.

In most cases, you will have 1 decoration, having multiple is less common.
In that case, if I have "Strikethrough", while doing "cheaper check first", I had to go through 8 branches while in my scenario, you went through 4 branches (Math doesn't add up? Keep reading.). That's 50% off!

If I had two, let's say "Overline" and "Strikethrough", I went with 2 branches in both scenarios for "Overline", and then I went with 7 branches while doing "cheaper check first"; and in my scenario, you went with 7 branches too while the first one was more expensive. But did we go with 7 branches in my scenario and was the first one more expensive? (That's why I said keep reading)

"Overline" and "Baseline" have the same length = 8; which means that if the evaluated input is of different length (in this case it is = "Strikethrough"), we do not compare the string itself, and by putting a string comparison first, we can eliminate check for "Baseline" if "Overline" hasn't matched the length check already. Suddenly my scenario wins 4 to 7 branches. Yet again, looks better for my scenario.

And if it's "Overline" and "Baseline"? Well "Overline" took 2 branches, that's constant, "Baseline" then one took 3 branches also in both scenarios and mine was slower because we had to compare "Baseline" to "Overline". That is the only edge case. This will fit into single vector register and I'm ready to sacrifice this one edge-case.

Comparing on a general purpose register if we don't have spills (compiler should optimize in a loop like this, and JIT will) is the same for length and bits here.

Reply:
By the way, a quick search on Github will show almost 6.7k `Underline` and 450 `Strikethrough`, nobody uses `Baseline` and `Overline`. It makes sense tbh and matches reality imho. I don't see much use for the two.

So the improvement that would make sense here is to put `Underline` first and `Strikethrough` second but that hurts the estetics of the code, haha. So I'm gonna keep it as it is tbh.

Reply:
I understand your point and ran a minimal benchmark locally to double check. In my benchmark the results were not consistent but with both of my suggestion (https://github.com/dotnet/wpf/pull/9778/files#r1759640738 and https://github.com/dotnet/wpf/pull/9778/files#r1759640864) it tends to be a smidge faster (I tested with "Strikethrough" and "Strikethrough,Underline").

I really think you should do my suggestion [here](https://github.com/dotnet/wpf/pull/9778/files#r1759640738) because it makes the code cleaner.

But for this suggestion I don't care, we're in the single digit nanosecond range which will probably never matter in any real world scenario. I still think it would be better and that the readability would be better but I'm not going to fight you on it.

Reply:
I do not know what kind of minimal benchmark you ran, I've analyzed the generated assembly previously just to be sure I still remember things correctly and then I wrote my points. If you ran the whole `ConvertToString` function, then of course your benchmarks are not conclusive.

Today I ran the benchmark of the loop with the essentials (e.g. just flag/bool setting), with input of "Strikethrough, Underline", both tiered and non-tiered compilation, 15 times, checked the hardware counters for instructions executed and it is always conclusive, not a single false positive, my solution is always 1-2 ns faster.

I've retested it both on Intel and AMD silicon (haha, getting BenchmarkDotNet to retrieve the hardware counters from ETW without modifying the source is some work as it relies on constants and at least in Win10 they're named differently), everytime with AVX2 (not AVX512 because it never matters here).

Which is obvious, besides the additional branching with this suggestion, the bool suggestion needs to load and store for each boolean.

I do not understand how are bit flags for a programmer less readable than 4 separate bools hanging around; however I'm happy to abstract it — make it into a private enum and using `HasFlag` to check if you think it helps readibility.

Reply:
My benchmark was less isolated so it may have been more susceptible to noise. It's probably why I couldn't reproduce the 1-2 ns faster you speak of. Since my benchmarks were not conclusive and yours were, it's fine to keep your code as is.

> I do not understand how are bit flags for a programmer less readable than 4 separate bools hanging around; however I'm happy to abstract it — make it into a private enum and using HasFlag to check if you think it helps readibility.

That's fine, I just tend to prefer the code to be as simple as possible (Even if a couple of ns slower). Bit flags are still simple but IMO bool variables are simpler. It's fine to keep your PR as is if that's what you prefer.

I don't have any other comments about this PR.